flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobSubmitHandler.java

```diff
@@ -148,11 +150,30 @@ private CompletableFuture<JobGraph> loadJobGraph(
                                         HttpResponseStatus.BAD_REQUEST,
                                         e));
                     }
+                    try {
+                        applyParallelismOverrides(jobGraph);
+                    } catch (Exception e) {
+                        throw new CompletionException(
+                                new RestHandlerException(
+                                        "Failed to apply parallelism overrides",
+                                        HttpResponseStatus.INTERNAL_SERVER_ERROR,
+                                        e));
+                    }
                     return jobGraph;
                 },
                 executor);
     }
 
+    private void applyParallelismOverrides(JobGraph jobGraph) {
+        Map<String, Integer> overrides = configuration.get(PipelineOptions.PARALLELISM_OVERRIDES);
+        for (JobVertex vertex : jobGraph.getVertices()) {
+            Integer override = overrides.get(vertex.getID().toHexString());
+            if (override != null) {
+                vertex.setParallelism(override);
```

Comment:
These should be logged, at least on debug, as otherwise good look figuring out where this comes from.